packages/core/utils/src/modules-sdk/define-link.ts

```diff
@@ -0,0 +1,316 @@
+import { LinkModulesExtraFields, ModuleJoinerConfig } from "@medusajs/types"
+import { isObject, pluralize, toPascalCase } from "../common"
+import { composeLinkName } from "../link"
+
+type InputSource = {
+  serviceName: string
+  field: string
+  linkable: string
+  primaryKey: string
+}
+
+type InputToJson = {
+  toJSON: () => InputSource
+}
+
+type CombinedSource = Record<any, any> & InputToJson
+
+type InputOptions = {
+  source: CombinedSource | InputSource
+  isList?: boolean
+}
+
+type ExtraOptions = {
+  pk?: {
+    [key: string]: string
+  }
+  database?: {
+    table: string
+    idPrefix?: string
+    extraColumns?: LinkModulesExtraFields
+  }
+}
+
+type DefineLinkInputSource = InputSource | InputOptions | CombinedSource
+
+type ModuleLinkableKeyConfig = {
+  module: string
+  key: string
+  isList?: boolean
+  primaryKey: string
+  alias: string
+  shortcuts?: {
+    [key: string]: string | { path: string; isList?: boolean }
+  }
+}
+
+function isInputOptions(input: any): input is InputOptions {
+  return isObject(input) && "source" in input
+}
+
+function isInputSource(input: any): input is InputSource {
+  return (isObject(input) && "serviceName" in input) || "toJSON" in input
+}
+
+function isToJSON(input: any): input is InputToJson {
+  return isObject(input) && "toJSON" in input
+}
+
+export function defineLink(
+  leftService: DefineLinkInputSource,
+  rightService: DefineLinkInputSource,
+  linkServiceOptions?: ExtraOptions
+) {
+  let serviceAObj = {} as ModuleLinkableKeyConfig
+  let serviceBObj = {} as ModuleLinkableKeyConfig
+
+  if (isInputSource(leftService)) {
+    const source = isToJSON(leftService) ? leftService.toJSON() : leftService
+
+    serviceAObj = {
+      key: source.linkable,
+      alias: source.field,
+      primaryKey: source.primaryKey,
+      isList: false,
+      module: source.serviceName,
+    }
+  } else if (isInputOptions(leftService)) {
+    const source = isToJSON(leftService.source)
+      ? leftService.source.toJSON()
+      : leftService.source
+
+    serviceAObj = {
+      key: source.linkable,
+      alias: source.field,
+      primaryKey: source.primaryKey,
+      isList: leftService.isList ?? false,
+      module: source.serviceName,
+    }
+  } else {
+    throw new Error("Invalid value for serviceA config")
```

Comment:
Could we do something like this in the error messages?
```suggestion
    throw new Error("Invalid link passed for ${module name} config")
```

Reply:
if it is neither of the previous check, then we dont have the module name, should I say first args and second for the other onw